### PR TITLE
Improve antichain inclusion and refactor code (renaming, suggestions)

### PR DESCRIPTION
**Pull request by kilohsakul**
- actualised antichain_inclusion to use sync. iterator for computing post,
- renamed few things, mainly incl -> inclusion
- some TODO comments in the code (maybe not needed to merge all of them)

### CODE REVIEW

**Comment by martinhruska** [comment]
Firstly, we should do reviews, fix the arised conflict and fix python-binding.

File: AUTOMATAFORMAT.md [hunk 1]
```diff
@@ -37,7 +37,7 @@ We use the general form of AFA, by Pavol V., and simple NFA.
   * By **enumeration**.
   * By **type-markers**. The literal is preceeded by a single character that specifies the type, we use `q` for states, `a` for symbols, and `n` for shared nodes. The marker *is* a part of the thing's name. 
   * **Automatic**. If all the other types are specified by enumeration or by markers, the third type may be marked as automatic, the parser will assign it to everything which is not of the two previous types. 
-* The typing scheme is specified as a key value line with the key  `<Thing>-<Scheme>` where  `<Thing>` may be `Alphabet`, `States`, or `Nodes`. `<Scheme>` may be `marked`, `enum`, or `auto`.  The scheme `enum` is followed by an enumeration. For instance, we can have `%Alphabet-enum a b c karel`, `%Alphabet-auto`, `%Alphabet-marked`.
+* The typing scheme is specified as a key value line with the key  `<Number>-<Scheme>` where  `<Number>` may be `Alphabet`, `States`, or `Nodes`. `<Scheme>` may be `marked`, `enum`, or `auto`.  The scheme `enum` is followed by an enumeration. For instance, we can have `%Alphabet-enum a b c karel`, `%Alphabet-auto`, `%Alphabet-marked`.
```
Comment (martinhruska) [inline AUTOMATAFORMAT.md:40]: why is this rename? An accident with auto replace in CLion?

File: include/mata/nfa.hh [hunk 2]
```diff
@@ -70,6 +58,9 @@ struct Run {
 using StringToStateMap = std::unordered_map<std::string, State>;
 using StringToSymbolMap = std::unordered_map<std::string, Symbol>;
 
+
+
+
```
Comment (martinhruska) [inline include/mata/nfa.hh:63]: please remove empty lines.

File: include/mata/nfa.hh [hunk 3]
```diff
@@ -298,12 +288,12 @@ struct Move {
...
 /// List of transitions from a certain state. Each element holds transitions with a certain symbol.
-using Moves = Mata::Util::OrdVector<Move>;
+        using Moves = Mata::Util::OrdVector<Move>;
 /// Transition relation for an NFA. Each index 'i' to the vector represents a state 'i' in the automaton.
-using TransitionRelation = std::vector<Moves>;
+        using TransitionRelation = std::vector<Moves>;
 
 /// An epsilon symbol which is now defined as the maximal value of data type used for symbols.
-constexpr Symbol EPSILON = limits.maxSymbol;
+        constexpr Symbol EPSILON = limits.maxSymbol;
```
Comment (martinhruska) [inline include/mata/nfa.hh:296]: Were the three changes above intentional or just accidental change of indent?

File: include/mata/nfa.hh [hunk 4]
```diff
@@ -357,6 +350,8 @@ public:
 
     auto states_number() const { return transition_relation.size(); }
 
+    //TODO: why this? Maybe we could have add_state(int state)
+    //Btw, why do we need adding states actually, we could just add states with transitions, or with initial/final states.
```
Comment (martinhruska) [inline include/mata/nfa.hh:354]: It comes from nature of Vector. It is more effecient to do resize if you know that you will need more states than adding one item in time.

File: include/mata/nfa.hh [hunk 5]
```diff
@@ -561,6 +557,7 @@ public:
      * @param states_to Set of target states.
      */
     void add_trans(State state_from, Symbol symbol, const StateSet& states_to);
+    //TODO: rename all "trans" to "transition". At lest function names.
```
Comment (martinhruska) [inline include/mata/nfa.hh:560]: disagree.

File: include/mata/synchronized-iterator.hh [hunk 6]
```diff
@@ -5,6 +5,7 @@
 #ifndef LIBMATA_SYNCHRONIZED_ITERATOR_HH
 #define LIBMATA_SYNCHRONIZED_ITERATOR_HH
 
+
```
Comment (martinhruska) [inline include/mata/synchronized-iterator.hh:8]: remove empty line.

File: include/mata/util.hh [hunk 7]
```diff
@@ -19,6 +19,7 @@
 #ifndef _MATA_UTIL_HH_
 #define _MATA_UTIL_HH_
 
+#include <bitset>
```
Comment (martinhruska) [inline include/mata/util.hh:22]: why is this import here when nothing else changed in this file? Only empty lines were added.

File: src/nfa/nfa-inclusion.cc [hunk 8]
```diff
@@ -225,10 +281,11 @@ bool Mata::Nfa::is_incl(
...
 	return algo(smaller, bigger, alphabet, cex, params);
-} // is_incl }}}
+} // is_included }}}
 
 bool Mata::Nfa::are_equivalent(const Nfa& lhs, const Nfa& rhs, const Alphabet *alphabet, const StringMap& params)
 {
+    //TODO: add comment on what this is doing, what is __func__ ...
     AlgoType algo{ set_algorithm(std::to_string(__func__), params) };
 
     if (params.at("algo") == "naive") {
```
Comment (martinhruska) [inline src/nfa/nfa-inclusion.cc:291]: Indentation in this file is a big mess. You don't see it because your editor solve it by you but check it here https://github.com/VeriFIT/mata/blob/c6c21c11bc7263b58285f1879ed42a3b1ec695cc/src/nfa/nfa-inclusion.cc
Anyone else would see it like this and it complicates orientation in source file.

File: include/mata/nfa.hh (hunk 5 above)
Comment (tfiedor) [inline include/mata/nfa.hh:560]: I agree, it's 2022, not middle ages.

File: include/mata/synchronized-iterator.hh [hunk 9]
```diff
@@ -173,7 +174,19 @@ namespace Mata {
...
+            Iterator next_minimum; // the value we should synchronise on after the first next call of advance().
+
+            bool is_synchronized() {
+                return !currently_synchronized.empty();
+            }
+
+            const Iterator get_current_minimum()
+            {
+                if (currently_synchronized.empty())
+                    throw std::runtime_error("Trying to get minimum from sync. ex. iterator which has no minium. Don't do that ever again or your nose falls off!");
```
Comment (tfiedor) [inline include/mata/synchronized-iterator.hh:187]: Fix typo (minium). Other than that, I really like it ;)

File: src/nfa/nfa-inclusion.cc [hunk 10]
```diff
@@ -152,25 +188,44 @@ bool Mata::Nfa::Algorithms::is_incl_antichains(
...
-							ds->erase(to_remove);
-						} else {
-							++it;
-						}
-					}
-
-					// TODO: set pushing strategy
-					ds->push_back(succ);
-				}
+                // prune data structures and insert succ inside
```
Comment (tfiedor) [inline src/nfa/nfa-inclusion.cc:191]: I would not keep commented code in here. Delete or maybe add guard? `#if ON_MACOS ... #endif`?

File: src/nfa/nfa-inclusion.cc [hunk 11]
```diff
@@ -152,25 +188,44 @@ bool Mata::Nfa::Algorithms::is_incl_antichains(
...
+                //            ++it;
+                //        }
+                //    }
+
+                //    // TODO: set pushing strategy
+                //    ds->push_back(succ);
+                //}
+
+                for (std::deque<ProdStateType>* ds : {&processed, &worklist}) {
+                    for (size_t it = 0; it < ds->size(); ++it) {
```
Comment (tfiedor) [inline src/nfa/nfa-inclusion.cc:212]: As usual: extract the `size()` if it is possible.

File: src/nfa/nfa-inclusion.cc [hunk 12]
```diff
@@ -152,25 +188,44 @@ bool Mata::Nfa::Algorithms::is_incl_antichains(
...
+                        }
+                    }
+
+                    // TODO: set pushing strategy
+                    ds->push_back(succ);
+                }
 
 				// also set that succ was accessed from state
-				paths[succ] = {prod_state, symb};
+				paths[succ] = {prod_state, smaller_symbol};
```
Comment (tfiedor) [inline src/nfa/nfa-inclusion.cc:228]: Maybe rename `prod_state` to `product_state` as well? I don't like shortenings.

File: src/nfa/nfa-inclusion.cc [hunk 13]
```diff
@@ -225,10 +281,11 @@ bool Mata::Nfa::is_incl(
 { // {{{
     AlgoType algo{ set_algorithm(std::to_string(__func__), params) };
 	return algo(smaller, bigger, alphabet, cex, params);
-} // is_incl }}}
+} // is_included }}}
 
 bool Mata::Nfa::are_equivalent(const Nfa& lhs, const Nfa& rhs, const Alphabet *alphabet, const StringMap& params)
 {
+    //TODO: add comment on what this is doing, what is __func__ ...
```
Comment (tfiedor) [inline src/nfa/nfa-inclusion.cc:288]: `__func__` is afaik preprocessor "constant" that is translated to the name of the current function (similar to `__line__` (current line in code) or `__file__` (current file).

**Review by tfiedor — APPROVED** [review]
I'm OK with the PR, there are some useless whitelines IMO, that could be removed, and lots of TODOs, that seems scary. Maybe we could destroy them as soon as possible.

File: src/nfa/nfa-inclusion.cc (hunk 8 above)
Comment (tfiedor) [reply src/nfa/nfa-inclusion.cc:291]: Yes, I agree. I recommend to never use tabs and some kind of "autotranslation" of tabs to hard spaces.

File: AUTOMATAFORMAT.md (hunk 1 above)
Comment (kilohsakul) [reply AUTOMATAFORMAT.md:40]: jj

File: include/mata/inter-aut.hh [hunk 14]
```diff
@@ -1,3 +1,4 @@
+//TODO: rename to intermediate_afa / afa-intermediate, or something like that.
```
[the remark this answers is not in the document]
Comment (kilohsakul) [reply include/mata/inter-aut.hh:1]: It actually is a kind of afa. Anyway, I somehow don't like those abbreviated names. 

Another thing about names of files, why do we have the prefix nfa or afa? They could be dropped no?

File: include/mata/nfa.hh (hunk 2 above)
Comment (kilohsakul) [reply include/mata/nfa.hh:63]: ok

File: include/mata/nfa.hh (hunk 3 above)
Comment (kilohsakul) [reply include/mata/nfa.hh:296]: probably, no ideal how it happened

File: include/mata/nfa.hh (hunk 5 above)
Comment (kilohsakul) [reply include/mata/nfa.hh:560]: argh, at least the names of functions?

File: include/mata/util.hh (hunk 7 above)
Comment (kilohsakul) [reply include/mata/util.hh:22]: it was some experiment, will remove

File: src/nfa/nfa-inclusion.cc (hunk 8 above)
Comment (kilohsakul) [reply src/nfa/nfa-inclusion.cc:272]: ok

File: include/mata/nfa.hh [hunk 15]
```diff
@@ -399,6 +394,7 @@ public:
         for (const State& st : vec) { this->make_initial(st); }
     }
 
+    // TODO: aaargh, the name should be is_initial, no? And we should probably implement it differently, as a boolean flag, to make this constant time.
```
[the remark this answers is not in the document]
Comment (kilohsakul) [reply include/mata/nfa.hh:397]: seems that this will be anyway replaced by something initial[q]

File: include/mata/nfa.hh (hunk 5 above)
Comment (kilohsakul) [reply include/mata/nfa.hh:560]: I think of myslelf as a knight saving abreviated names in distress.

File: src/nfa/nfa-inclusion.cc (hunk 11 above)
Comment (kilohsakul) [reply src/nfa/nfa-inclusion.cc:193]: not my foult, ask Ondra :)

Comment (kilohsakul) [reply src/nfa/nfa-inclusion.cc:193]: Aha but I added my favourite trick so that size changes during the iteration :).

File: src/nfa/nfa-inclusion.cc (hunk 10 above)
Comment (kilohsakul) [reply src/nfa/nfa-inclusion.cc]: should be removed.

**Comment by kilohsakul** [comment]
reviewer pleaseing done

File: include/mata/nfa.hh (hunk 5 above)
Comment (martinhruska) [reply include/mata/nfa.hh:556]: I think one should pay some bucks for characters wasted in naming of things.

File: include/mata/inter-aut.hh (hunk 14 above)
[the remark this answers is not in the document]
Comment (martinhruska) [reply include/mata/inter-aut.hh:1]: Yes, it is a kind of afa because nfa is kind of afa. However, a spirit of this class is to be an intermediate representation between parser and efficient automata data structures. Therefore relating it to a certain kind of automata would be confusing.

**Comment by martinhruska** [comment]
@tfiedor are you satisfied with changes?

**Comment by tfiedor** [comment]
> @tfiedor are you satisfied with changes?

I trust you and Mr. L. that all that needed to be done was done.
